### PR TITLE
Implemented isolated context for Fly plugins

### PR DESCRIPTION
**Pull request by hzlmn**
Basically it isolate each plugin with **shallow** copy of **fly** instance, 
then it get differences of original **fly** instance and context that was used by plugins. 

Then safely inject changes, and if changes was produced on sealed prop (like root, because we dont want anyone to rewrite this context) it's **throws** an _exception_

@lukeed @jbucaran

### CODE REVIEW

**Comment by hzlmn** [comment]
Seems like there are some minor issues with `tape` like tests without end and plan count missmatch. Should I fix it?

**Comment by lukeed** [comment]
@hzlmn The non-plugin tests should still be passing. I just did a fresh clone & install and the plans & counts are fine.

What it probably signifies is that your promises aren't resolving properly, and so the test never "knows" it's finished.

File: lib/utils/array-to-map.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+module.exports = arrayToMap
+
+/**
+ * Create Map object from single list
+ *
+ * @param  {Array} list
+ * @return {Object}
+ */
+function arrayToMap(list) {
```
Comment (lukeed) [inline lib/utils/array-to-map.js:9]: lol, i think someone works with Elixir... @hzlmn

**Comment by hzlmn** [comment]
@lukeed What promises do you mean?

**Comment by hzlmn** [comment]
@lukeed My bad, I added `t.end()` in async tests,  and that is why `tape` skipped async tests.

File: lib/utils/array-to-map.js (hunk 1 above)
Comment (hzlmn) [reply lib/utils/array-to-map.js:9]: Not very much in it, but yes. Btw indexing map is way more faster than array.

**Comment by hzlmn** [comment]
I am closing this. I gonna submit new one a bit cleaner and with some improvements

**Comment by lukeed** [comment]
@hzlmn Hey sorry, I didn't have a whole lot of spare time to review it. I'll be better about your next one

**Pull request closed without merging**